internal/helm/helmadapter/helm3_env_service.go

```diff
@@ -420,16 +421,23 @@ func (h helm3EnvService) listCharts(_ context.Context, helmEnv helm.HelmEnv, fil
 
 // getDetailedChart gets the chart details from the chart archive
 func (h helm3EnvService) getDetailedCharts(_ context.Context, helmEnv helm.HelmEnv, repoVersions repo.ChartVersions) (map[string]*chart.Chart, error) {
-	// set up a "fake" chart repo to use it's getter capabilities
-	chartRepo, err := repo.NewChartRepository(&repo.Entry{URL: "http://test"}, getter.All(h.processEnvSettings(helmEnv)))
-	if err != nil {
-		return nil, errors.WrapIf(err, "failed to setup downloader")
-	}
+	getters := getter.All(h.processEnvSettings(helmEnv))
 
 	detailedCharts := make(map[string]*chart.Chart)
 	for _, repoChartVersionPtr := range repoVersions {
 		// todo check the other urls, other checks?
-		buffer, err := chartRepo.Client.Get(repoChartVersionPtr.URLs[0])
+		repoUrl := repoChartVersionPtr.URLs[0]
+		u, err := url.Parse(repoUrl)
+		if err != nil {
+			return nil, errors.Errorf("invalid chart URL format: %s", repoUrl)
+		}
+
+		client, err := getters.ByScheme(u.Scheme)
+		if err != nil {
+			return nil, errors.Errorf("could not find protocol handler for: %s", u.Scheme)
+		}
+
+		buffer, err := client.Get(repoUrl)
```

Comment:
**Required**: this solution only works if the Helm S3 repository is in the same region where the AWS credential context is.

0. I have everything configured locally to `eu-central-1` by default (AWS credentials, Pipeline secret, everything).

1. I created a Helm S3 repo in us-west-1 and uploaded the NPLS chart there according to the Helm S3 guide.

2. I added the following to my local Pipeline control-plane configuration:
```yaml
cluster:
    labels:
        domain: banzaicloud.io
        forbiddenDomains:
            [
                "k8s.io",
                "kubernetes.io",
                "google.com",
                "coreos.com",
                "oraclecloud.com",
                "node.info",
                "azure.com",
                "agentpool",
                "storageprofile",
                "storagetier",
            ]

        charts:
            nodepoolLabelOperator:
                chart: "s3/nodepool-labels-operator"
                version: "0.1.1"

helm:
    home: "./var/cache"
    repositories:
        stable: "https://charts.helm.sh/stable"
        banzaicloud-stable: "https://kubernetes-charts.banzaicloud.com"
        bitnami: "https://charts.bitnami.com/bitnami"
        loki: "https://grafana.github.io/loki/charts"
        prometheus-community: "https://prometheus-community.github.io/helm-charts"
        s3: "s3://my-bucket-from-us-west-1/stable/charts"
```

3. I started a simple cluster creation with this configuration:
```yaml
{
  "cloud": "amazon",
  "location": "eu-central-1",
  "name": "pregnor-pipeline-helm-s3-test",
  "properties": {
    "eks": {
      "nodePools": {
        "pool1": {
          "autoscaling": true,
          "count": 3,
          "instanceType": "t3a.medium",
          "maxCount": 10,
          "minCount": 1,
          "tags": {
            "nodePoolTag": "NodePoolTagValue"
          }
        }
      },
      "tags": {
        "clusterTag": "clusterTagValue"
      },
      "version": "1.18"
    }
  },
  "secretName": "my-secret"
}
```

4. The cluster creation failed at the NPLS installation step.
```shell
ERRO[1577] Activity error.
    ActivityType=install-nodepool-labelset-operator
    Domain=pipeline RunID=3c765d83-62cb-4d72-870c-b3e1fae24b5f 
    TaskList=pipeline 
    WorkerID=14912@localhost@pipeline 
    WorkflowID=942f1e6c-f037-4943-b60b-309b02f3e127_2 
    application=pipeline.worker 
    component=cadence-worker 
    environment=production 
    error="installing NodePoolLabelSet operator failed: failed to install release: failed to locate chart: failed to download \"s3/nodepool-labels-operator\" at version \"0.1.1\" (hint: running `helm repo update` may help)" errorVerbose="failed to download \"s3/nodepool-labels-operator\" at version \"0.1.1\" (hint: running `helm repo update` may help)
helm.sh/helm/v3/pkg/action.(*ChartPathOptions).LocateChart
	_build/helm.sh/helm/v3/pkg/action/install.go:690
github.com/banzaicloud/pipeline/internal/helm/helmadapter.releaser.Install
	internal/helm/helmadapter/releaser_service.go:99
github.com/banzaicloud/pipeline/internal/helm.service.InstallRelease
	internal/helm/service.go:470
github.com/banzaicloud/pipeline/internal/helm/helmadapter.(*helm3UnifiedReleaser).InstallDeployment
	internal/helm/helmadapter/helm_service.go:168
github.com/banzaicloud/pipeline/internal/cluster/clustersetup.InstallNodePoolLabelSetOperatorActivity.Execute
	internal/cluster/clustersetup/activity_install_nodepool_labelset_operator.go:67
reflect.Value.call
	/usr/local/Cellar/go/1.16.3/libexec/src/reflect/value.go:476
reflect.Value.Call
	/usr/local/Cellar/go/1.16.3/libexec/src/reflect/value.go:337
go.uber.org/cadence/internal.(*activityExecutor).Execute
	_build/go.uber.org/cadence/internal/internal_worker.go:731
go.uber.org/cadence/internal.(*activityTaskHandlerImpl).Execute
	_build/go.uber.org/cadence/internal/internal_task_handlers.go:1822
go.uber.org/cadence/internal.(*activityTaskPoller).ProcessTask
	_build/go.uber.org/cadence/internal/internal_task_pollers.go:1014
go.uber.org/cadence/internal.(*baseWorker).processTask
	_build/go.uber.org/cadence/internal/internal_worker_base.go:323
runtime.goexit
	/usr/local/Cellar/go/1.16.3/libexec/src/runtime/asm_amd64.s:1371
failed to locate chart
failed to install release
installing NodePoolLabelSet operator failed"
```

5. I created a Helm S3 repo in eu-central-1 and uploaded the NPLS chart there too.

6. I reconfigured my Pipeline control-plane to use the eu-central-1 Helm S3 repo for the NPLS chart and restarted both the API server and the worker.

7. The worker successfully installed the NPLS on the managed cluster afterwards and the cluster creation finished successfully.

For more information see the [plugin's documentation](https://github.com/hypnoglow/helm-s3#s3-bucket-location).

There are 2 ways to resolve this: either we have to wait for the dynamic bucket region PR to be merged and released in the original helm-s3 plugin or we have to use our fork of the plugin here as well as in the installer which is still WIP (trying to finish it ASAP).

Reply:
Fixed in d894661.